### ISSUE TITLE
`jbang init` has `{baseName}` ambiguities in templates

### ISSUE THREAD

**Issue by quintesse** [jbangdev/jbang#987]
See discussion here https://github.com/jbangdev/jbang-catalog/pull/8#discussion_r702797278

The TL;DR is that using any complex file names using `{baseName}` cause problems.

The longer explanation is that when file names and/or class names are generated that are not plain `{baseName}` but have some string prepended or appended, eg `{baseName}Test` are problematic because `{baseName}` can have different values depending on context, it's just that in the simple cases we've used so far the `{baseName}` just happened to resolve to the same value.

For example when you write `jbang init -t foo Foo.java` the value of `{baseName}` used in the `jbang-catalog.json` file resolves to "Foo". So if that catalog file has a section like this:

```
"file-refs": {
    "{basename}Test.java": "templates/junit4b.java.qute"
},
```

it means that jbang will create a file named "FooTest.java" with it's contents based on `templates/junit4b.java.qute`, but... and this is the issue, once that template gets processed the `{baseName}` has the value "FooTest" because that's what its name is, so that's correct too.

**Comment by quintesse** [comment]
IMO the problem is that we conflated to different kinds of value: the name that the user passed on the command line and the name of the file being processed. So this simplest solution would be to use different names, right?

Perhaps something like `scriptName` and `scriptBasename` vs `fileName` and `fileBasename`? (in case of the jbang catalog file only the first two would be supported of course)

**Comment by quintesse** [comment]
Current consensus is to use `initFile`and `initBasename`.

(Also: check if those names are case-sensitive and if so make them -insensitive)

**Comment by quintesse** [comment]
Ok, I just realized that in our code "basename" means, the name we'll use for the class inside the script file.
Which is of course very different from what's normally understood as a file's basename.
Still, in templates both values can be useful...

So I'm wondering if we won't need an even more extensive list of names, something like:

- `initFile` - the original script reference / file name (no path)
- `initBaseName` - the original name without extension
- `initClassName` - what we call "basename", a CamelCased version of the original name (without extension)
- `fileName` - the file name of the current template being processed
- `baseName` - that same name without extension
- `className` - a CamelCased version of the template file name (without extension)

Dunno, wdyt @maxandersen ?

**Comment by maxandersen** [comment]
`basename` is camelcased based on the base name of the file - I don't think we can change that.

We can add more explicit ones, something like if you do `jbang init -t sometemplate src/testing.java`
you would have:


- `initFile` - the original script reference / file name (no path): `testing.java`
- `initBaseName` - the original name without extension: `testing`
- `initClassName` and `baseName` - a CamelCased version of the original name (without extension): `Testing`
- `templateFile` - the file name of the current template being processed: `abc.qute.java`
- `tempateBaseName` - that same name without extension: `abc`
- `className` - a CamelCased version of the template file name (without extension): ...not understanding this one ?

btw. why not use `scriptRef` instead of `init*` ?

**Comment by quintesse** [comment]
> btw. why not use scriptRef instead of init* ?

What? You just agreed on `init*` yourself only a few days ago. :-) ☝️ https://github.com/jbangdev/jbang/issues/987#issuecomment-916950109

**Comment by quintesse** [comment]
About the template file, a couple of comments:

1. "basename is camelcased based on the base name of the file - I don't think we can change that.", why not?
2. `abc.qute.java` I guess you meant `abc.java.qute`, right?
3. "...not understanding this one", in your example this would be `Abc`.
4. If `templateFile` is `abc.java.qute` I don't think `templateBasename` can be `abc`, it should be `abc.java`. But that doesn't make much sense. So perhaps `templateBaseName` should be `abc` but then `templateFile` should be `abc.java`.
5. But point 4 still doesn't make much sense to me, so perhaps for template files we actually need 3 names? One for the original template file (the one with the `.qute` extension), one for the resulting file (`abc.java`) and one for the base name of the resulting file, like this for example:

   - `templateFile` - the original template file name `abc.java.qute`
   - `targetFile` - the result of the templating operation `abc.java`
   - `targetBaseName` - the same name without extension `abc`

**Comment by quintesse** [comment]
This is a *very* long message that I just used to gather ideas. No need to actually read it 😉 

A full example using names that started this whole oddysee. Imagine wanting to create a JUnit test file for a class called `SomeClass`:

```
jbang init -t junit SomeClass
```

This would give:

- `initFile` - would this have a value in this case? Perhaps just simply `SomeClass` or nothing?
- `initBaseName` - same here `SomeClass` or nothing?
- `initClassName` and `baseName` - `SomeClass`
- `templateFile` - something like `junit.java.qute`
- `targetFile` - the template uses `{basename}Test.java`, so `SomeClassTest.java`
- `targetBaseName` - that same name without extension, so `SomeClassTest`
- `className` - `SomeClassTest`

But if you wrote:

```
jbang init -t junit tests/SomeClass.java
```

It would give:

- `initFile` - `SomeClass.java`
- `initBaseName` - `SomeClass`
- `initClassName` and `baseName` - `SomeClass`
- `templateFile` - something like `junit.java.qute`
- `targetFile` - the template uses `{basename}Test.java`, so `SomeClassTest.java`
- `targetBaseName` - that same name without extension, so `SomeClassTest`
- `className` - `SomeClassTest`

Ok, that does seem to give a lot of duplicates, so perhaps we can strip most of them. This set would be enough it seems:

- `scriptRef` - `SomeClass.java`
- `baseName` - `SomeClass`
- `templateFile` - something like `junit.java.qute`
- `targetFile` - the template uses `{basename}Test.java`, so `SomeClassTest.java`
- `targetBaseName` - that same name without extension, so `SomeClassTest`

Also not sure if we'd ever need `templateFile`...

**Comment by quintesse** [comment]
TL;DR

- `scriptRef` - could be something like `SomeClass`, `SomeClass.java` or `tests/SomeClass.java`
- `baseName` - `SomeClass`
- `targetFile` - the template uses `{basename}Test.java`, so `SomeClassTest.java`
- `targetBaseName` - that same name without extension, so `SomeClassTest`

**Comment by quintesse** [comment]
Forgive me the fact that I'm incoherently writing everything down without first giving things a proper thought :-)

But I just realized that perhaps there's just a _single_ change we need to make. We need to define that:

- `{baseName}` is the camel cased version of `scriptRef`


Longer explanation:

Thing is that all other values can be determined from that `baseName` as long as it's based on the original `scriptRef` and not the name of the file being processed. Because if the file being processed is called `abc.java` the dev already knows it's name, no need to add define any variables. If the name is `{baseName}Test.java` then you can just keep using `{baseName}Test` throughout the code. THe whole problem was caused by the fact the `baseName` took on different values depending on where it was used, and that's the only thing that needs fixing.

**Comment by maxandersen** [comment]
makes sense :)
